url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/packer/MessagePackPacker.java

```diff
@@ -0,0 +1,334 @@
+        if (d < -(1L << 5)) {
+            if (d < -(1L << 15)) {
+                if (d < -(1L << 31)) {
+                    // signed 64
+                    out.writeByteAndLong((byte) 0xd3, d);
+                } else {
+                    // signed 32
+                    out.writeByteAndInt((byte) 0xd2, (int) d);
+                }
+            } else {
+                if (d < -(1 << 7)) {
+                    // signed 16
+                    out.writeByteAndShort((byte) 0xd1, (short) d);
+                } else {
+                    // signed 8
+                    out.writeByteAndByte((byte) 0xd0, (byte) d);
+                }
+            }
+        } else if (d < (1 << 7)) {
+            // fixnum
+            out.writeByte((byte) d);
+        } else {
+            if (d < (1L << 16)) {
+                if (d < (1 << 8)) {
+                    // unsigned 8
+                    out.writeByteAndByte((byte) 0xcc, (byte) d);
+                } else {
+                    // unsigned 16
+                    out.writeByteAndShort((byte) 0xcd, (short) d);
+                }
+            } else {
+                if (d < (1L << 32)) {
+                    // unsigned 32
+                    out.writeByteAndInt((byte) 0xce, (int) d);
+                } else {
+                    // unsigned 64
+                    out.writeByteAndLong((byte) 0xcf, d);
+                }
+            }
+        }
+        stack.reduceCount();
+    }
+
+    
+    protected void writeBigInteger(BigInteger d) throws IOException {
+        if (d.bitLength() <= 63) {
+            writeLong(d.longValue());
+            stack.reduceCount();
+        } else if (d.bitLength() == 64 && d.signum() == 1) {
+            // unsigned 64
+            out.writeByteAndLong((byte) 0xcf, d.longValue());
+            stack.reduceCount();
+        } else {
+            throw new MessageTypeException(
+                    "MessagePack can't serialize BigInteger larger than (2^64)-1");
+        }
+    }
+
+    
+    protected void writeFloat(float d) throws IOException {
+        out.writeByteAndFloat((byte) 0xca, d);
+        stack.reduceCount();
+    }
+
+    
+    protected void writeDouble(double d) throws IOException {
+        out.writeByteAndDouble((byte) 0xcb, d);
+        stack.reduceCount();
+    }
+
+    
+    protected void writeBoolean(boolean d) throws IOException {
+        if (d) {
+            // true
+            out.writeByte((byte) 0xc3);
+        } else {
+            // false
+            out.writeByte((byte) 0xc2);
+        }
+        stack.reduceCount();
+    }
+
+    
+    protected void writeByteArray(byte[] b, int off, int len)
+            throws IOException {
+        if (len < 32) {
+            out.writeByte((byte) (0xa0 | len));
+        } else if (len < 65536) {
+            out.writeByteAndShort((byte) 0xda, (short) len);
+        } else {
+            out.writeByteAndInt((byte) 0xdb, len);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 197:203, 211:217, 250:257, 287:294.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.